### ISSUE TITLE
Several Demos Show NaN for the loss Calculation only on Mobile

### ISSUE THREAD

**Issue by hpssjellis** [tensorflow/tfjs#90]
All these programs work fine on my desktop computer, but have a few issues on my Smartphone Samsung S8+ mobile device, running Android version 8.0  These 4 demos show NaN for the loss calculation:

Addition-RNN: Loss NaN then all other calculations are messed up. This demo does not work on my mobile device
[https://storage.googleapis.com/tfjs-examples/addition-rnn/dist/index.html](https://storage.googleapis.com/tfjs-examples/addition-rnn/dist/index.html)

Mnist-core: Only the Loss shows NaN when training, but demo works fine otherwise
[https://storage.googleapis.com/tfjs-examples/mnist-core/dist/index.html](https://storage.googleapis.com/tfjs-examples/mnist-core/dist/index.html)

Mnist Layers: Only Last Loss value shows NaN. Interesting that the Last Loss graph shows correct values.  The rest of program works fine
[https://storage.googleapis.com/tfjs-examples/mnist/dist/index.html](https://storage.googleapis.com/tfjs-examples/mnist/dist/index.html)

![last-loss-mnist-layers](https://user-images.githubusercontent.com/5605614/38293015-d16027ae-379a-11e8-9e63-849a2d9452c2.png)

Webcam Pacman training shows NaN and stops training. This demo does not work on my mobile device.

[https://storage.googleapis.com/tfjs-examples/webcam-transfer-learning/dist/index.html](https://storage.googleapis.com/tfjs-examples/webcam-transfer-learning/dist/index.html)

**Comment by hpssjellis** [comment]
Can someone else try on another Android 8.0 device, as the above demos work fine on an Android 7.1.1 tablet

**Comment by hpssjellis** [comment]
Here is the easiest demo I can make of the problem.


https://hpssjellis.github.io/beginner-tensorflowjs-examples-in-javascript/beginner-examples/android-issue.html


.

![android-issue2](https://user-images.githubusercontent.com/5605614/39322747-b0d47b6e-493f-11e8-9dbd-9a08dfda3439.png)









I just noticed that if I change the loss from 

`loss: 'meanSquaredError',`

to 
`loss: 'hinge',    `

I don't get NaN?

**Comment by lastmjs** [comment]
I'm getting similar issues, I'm more or less running this demo: https://js.tensorflow.org/tutorials/webcam-transfer-learning.html and I keep getting NaN for most of my losses. It works fine on my desktop, but on my Nexus 5x running Android 8+ I get NaN.

**Comment by Saxenauts** [comment]
I'd like to study this. I am new to tfjs, so any pointers would be appreciated.

**Comment by hpssjellis** [comment]
I think it is to do with a learning rate that is too high for android. Will try to get a working demo going.

**Comment by Saxenauts** [comment]
Okay, I'm more interested in how the learning rates affect the training on
different hardwares. How  exactly does processor being an Android change
things ?

**Comment by hpssjellis** [comment]
No idea what is different about Android, but here is a [changed Addition RNN](https://hpssjellis.github.io/beginner-tensorflowjs-examples-in-javascript/tf-examples/addition-rnn/easier/index.html), that allows you to adjust the learning rate. Needs a learning rate of below 0.03 to work on Android 8.0.0,


Image with the regular learning rate of 0.2

![android-bad](https://user-images.githubusercontent.com/5605614/44074291-ddc1ef0e-9f4c-11e8-8453-144fb8b53139.jpg)


.

Image with a learning rate of 0.03
![androidgood](https://user-images.githubusercontent.com/5605614/44074292-ddd958b0-9f4c-11e8-8ed8-dc32f40b2d09.jpg)


Even 0.03 eventually got NaN for the loss.
Note: The addition RNN demo is too slow on Android for me to get it to learn all the values.

**Comment by hpssjellis** [comment]
this can close. It is more to do with Android needs a lower training rate than on a desktop computer
